daemon/src/model/cfd.rs

```diff
@@ -2663,77 +2651,99 @@ mod tests {
     #[test]
     fn ensure_collaborative_settlement_takes_rollover_fees_into_account() {
         let quantity = Usd::new(dec!(10));
-        let leverage = Leverage::new(2).unwrap();
         let opening_price = Price::new(dec!(10000)).unwrap();
         let closing_price = Price::new(dec!(10000)).unwrap();
         let positive_funding_rate = dec!(0.0001);
 
-        assert_eq!(
-            calculate_long_margin(opening_price, quantity, leverage),
-            Amount::from_sat(50000)
-        );
-        assert_eq!(
-            calculate_short_margin(opening_price, quantity),
-            Amount::from_sat(100000)
-        );
+        // Expected payout at closing-price-interval defined by payout curve
+        let expected_payout_taker = Amount::from_sat(49583);
+        let expected_payout_maker = Amount::from_sat(100247);
+        // Expected initial funding fee based on the funding rate and short-margin (because the rate
+        // is positive meaning long pays short)
+        let expected_initial_funding_fee = Amount::from_sat(10);
+
+        // Note that transaction fees ARE zero for these tests because the order's fee-rate is set
+        // to 0!
+        //
+        // Note that the leverage is locked in at 2 because we create the cfd from the
+        // order
 
         // 0 rollover
         let (taker_payout, maker_payout) = collab_settlement_taker_long_maker_short(
             quantity,
-            leverage,
             opening_price,
             closing_price,
             positive_funding_rate,
-            1000,
             0,
+            0,
+        );
+        assert_eq!(
+            taker_payout,
+            expected_payout_taker - expected_initial_funding_fee
+        );
```

Comment:
Can we please split this test?

This would be much easier to read if these variables were declared right above the assertion. Esp. with the tests further down, having to scroll up and down to understand what these values are makes it really hard to understand.

Reply:
It would also make the tests a lot more concise because we could inline certain variables in the function call to `collab_settlement_taker_long_maker_short` and reserve the "arrange" part of the test for the values that you really care about. This test could for example be:

```rust
    #[test]
    fn given_collab_close_at_same_price_with_no_rollovers_pays_funding_fee_to_maker() {
        let opening_price = Price::new(dec!(10000)).unwrap();
        let closing_price = opening_price;
	let positive_funding_rate = dec!(0.0001);
	let num_rollovers = 0;

        let (taker_payout, maker_payout) = collab_settlement_taker_long_maker_short(
            Usd::new(dec!(10)),
            opening_price,
            closing_price,
            positive_funding_rate,
            0,
            num_rollovers,
        );

        let long_pays_short_funding_fee = 10;
        assert_eq!(taker_payout, Amount::from_sat(49583 - long_pays_short_funding_fee));
        assert_eq!(maker_payout, Amount::from_sat(100247 + long_pays_short_funding_fee));
}
```

Reply:
We could also make `collab_settlement_taker_long_maker_short` return a `u64` so we can remove the noise of creating `Amount`s.

Reply:
I still think that the test should be one because it is about regressions of rollover. But since this keeps bothering you more than me and it has been coming up twice I'll split them. The initial values will be duplicated in each test then.

Reply:
@bonomat @klochowicz do you have an opinion about this?

Reply:
> @bonomat @klochowicz do you have an opinion about this?

I don't have strong feelings about this.

Reply:
> The initial values will be duplicated in each test then.

There are ways to mitigate this :)

> But since this keeps bothering you more than me

It is less about bothering me or you. There are good reasons for it! Some background knowledge about testing:

The goal to always strive for with tests is:

- **one** statement for arrange
- **one** expression for act
- **one** statement for assert

Having more than one statement for arrange is fairly tolerable.
Having more than one expression for act is more smelly but tolerable if it just takes multiple (production code) calls to get something done. Can easily be mitigate though with something like you did above where you extract a helper for the test
Having more than one assert is tolerable IF they are a single block although ideally, we can assert the behaviour on a single outcome.

Multiple asserts per tests, intermixed with even more "act"s has numerous downsides:

1. An assert will early exit the test, meaning if a test fails (on CI for example) it is basically impossible to know, whether the other asserts would have passed. This is valuable information when debugging a problem to be able to narrow down, where the problem is. If you have one assert per test, you will learn things like: "Interesting, when doing no rollovers, things break but 2 rollovers are fine?".
2. An assert always states an expectation and a test is a form of documentation, how the system behaves. If you have more than one assert per test, it takes a good chunk focus and brainpower, to understand what part of the system is actually being tested here!

Reply:
I feel like we went on a tangent a bit with the tests:

As for tests, I personally find them often much harder to read after refactoring into the goal above - so, at least as the downside 2) I feel that my brainpower is fragmented in trying to simultaneously parse complicated setup functions & similar things scattered across different tests.
As with many other things, I perceive it as a trade-off - and I'd pick readability, speed & convenience vs a bit easier time debugging (the debugging can be quickly narrowed down if you re-run the test with debugger / stdout and look at it, so I don't think it's a big issue for us!)